Assertion failed during retyping of inlined call.

## Compiler version

3.0.0-RC2-bin-20210219-aa7c21e-NIGHTLY

When trying to extract small example for another bug report.

## Minimized code

X.scala:

```Scala
package x


import scala.annotation._
import scala.quoted._
import scala.compiletime._


trait CpsMonad[F[_]]:

  def pure[T](x:T):F[T]

  def impure[T](x:F[T]):T

  def map[A,B](x:F[A])(f: A=>B):F[B]


@compileTimeOnly("await should be inside async block")
def await[F[_],T](f:F[T])(using am:CpsMonad[F]):T = ???

inline given conversion[F[_],T](using CpsMonad[F]): Conversion[F[T],T] =
           x => await(x)


object X {

 inline def process[F[_], T](inline t:T)(using m: CpsMonad[F]):F[T] =
    ${ processImpl[F,T]('t, 'm) }


 def processImpl[F[_]:Type, T:Type](t:Expr[T], m:Expr[CpsMonad[F]])(using Quotes):Expr[F[T]] =
   import quotes.reflect._
   val r = processTree[F,T](t.asTerm, m.asTerm)
   r.asExprOf[F[T]]


 def processTree[F[_]:Type, T:Type](using Quotes)(t: quotes.reflect.Term, m: quotes.reflect.Term):quotes.reflect.Term =
   import quotes.reflect._
   val r: Term = t match
     case Inlined(_,List(),body) => processTree[F,T](body, m)
     case Inlined(d,bindings,body) =>
       Inlined(d,bindings,processTree[F,T](body, m))
     case Block(stats,expr) => Block(stats,processTree(expr, m))
     case Apply(Apply(TypeApply(Ident("await"),targs),List(body)),List(m)) => body
     case Apply(f,List(arg)) =>
          val nArg = processTree[F,String](arg, m)
          Apply(Apply(TypeApply(Select.unique(m,"map"),
                               List(Inferred(arg.tpe.widen),Inferred(t.tpe.widen))
                      ),
                      List(nArg)),
                 List(f)
          )
     case Apply(f,List()) =>
          Apply(TypeApply(Select.unique(m,"pure"),List(Inferred(t.tpe.widen))),List(t))
     case Typed(x,tp) => Typed(processTree(x,m), Inferred(TypeRepr.of[F].appliedTo(tp.tpe)) )
     case _ => throw new RuntimeException(s"tree not recoginized: $t")
   r


}
```
and Main.x:

```
package x

import scala.language.implicitConversions
import scala.concurrent.Future

given FutureAsyncMonad: CpsMonad[Future] with
  def pure[T](t:T): Future[T] = ???
  def impure[T](t:Future[T]): T = ???
  def map[A,B](x:Future[A])(f: A=>B): Future[B] = ???


object Api:

  def doSomething():Future[String] =
    Future.successful("doSomething")

  def println(x:String):Unit =
    Console.println(x)


object Main:

  def main(args: Array[String]): Unit =
    X.process[Future,Unit]{
       Api.println(Api.doSomething())
    }

```
## Output (click arrow to expand)
<details>

```scala
rssh@rssh4 wrong-compiletime-only % sbt compile                   
[info] welcome to sbt 1.4.5 (N/A Java 15.0.1)
[info] loading global plugins from /Users/rssh/.sbt/1.0/plugins
[info] loading settings for project wrong-compiletime-only-build from plugins.sbt ...
[info] loading project definition from /Users/rssh/tests/dotty/wrong-compiletime-only/project
Fetching latest Dotty nightly version...
Latest Dotty nightly build version: 3.0.0-RC2-bin-20210219-aa7c21e-NIGHTLY
[info] loading settings for project root from build.sbt ...
[info] set current project to test (in build file:/Users/rssh/tests/dotty/wrong-compiletime-only/)
[info] Executing in batch mode. For better performance use sbt's shell
[info] compiling 2 Scala sources to /Users/rssh/tests/dotty/wrong-compiletime-only/target/scala-3.0.0-RC2/classes ...
java.lang.AssertionError: assertion failed while compiling /Users/rssh/tests/dotty/wrong-compiletime-only/src/main/scala/x/Main.scala, /Users/rssh/tests/dotty/wrong-compiletime-only/src/main/scala/x/X.scala
[error] ## Exception when compiling 2 sources to /Users/rssh/tests/dotty/wrong-compiletime-only/target/scala-3.0.0-RC2/classes
[error] java.lang.AssertionError: assertion failed
[error] scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:11)
[error] dotty.tools.dotc.typer.EtaExpansion$.etaExpand(EtaExpansion.scala:226)
[error] dotty.tools.dotc.typer.Typer.adaptNoArgsUnappliedMethod$2(Typer.scala:3383)
[error] dotty.tools.dotc.typer.Typer.adaptNoArgs$1(Typer.scala:3528)
[error] dotty.tools.dotc.typer.Typer.adapt1(Typer.scala:3751)
[error] dotty.tools.dotc.typer.Typer.adapt(Typer.scala:3058)
[error] dotty.tools.dotc.typer.ProtoTypes$FunProto.typedArg(ProtoTypes.scala:416)
[error] dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:802)
[error] dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:802)
[error] dotty.tools.dotc.typer.Applications$Application.addTyped$1(Applications.scala:528)
[error] dotty.tools.dotc.typer.Applications$Application.matchArgs(Applications.scala:593)
[error] dotty.tools.dotc.typer.Applications$Application.init(Applications.scala:431)
[error] dotty.tools.dotc.typer.Applications$TypedApply.<init>(Applications.scala:697)
[error] dotty.tools.dotc.typer.Applications$ApplyToUntyped.<init>(Applications.scala:801)
[error] dotty.tools.dotc.typer.Applications.ApplyTo(Applications.scala:1012)
[error] dotty.tools.dotc.typer.Applications.ApplyTo$(Applications.scala:301)
[error] dotty.tools.dotc.typer.Typer.ApplyTo(Typer.scala:102)
[error] dotty.tools.dotc.typer.Applications.simpleApply$1(Applications.scala:859)
[error] dotty.tools.dotc.typer.Applications.realApply$5$$anonfun$4(Applications.scala:938)
[error] dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:2852)
[error] dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:949)
[error] dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:987)
[error] dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:301)
[error] dotty.tools.dotc.typer.Inliner$InlineTyper.typedApply(Inliner.scala:1301)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2597)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2659)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
[error] dotty.tools.dotc.typer.ProtoTypes$FunProto.$anonfun$2(ProtoTypes.scala:414)
[error] dotty.tools.dotc.typer.ProtoTypes$FunProto.cacheTypedArg(ProtoTypes.scala:370)
[error] dotty.tools.dotc.typer.ProtoTypes$FunProto.typedArg(ProtoTypes.scala:415)
[error] dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:802)
[error] dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:802)
[error] dotty.tools.dotc.typer.Applications$Application.addTyped$1(Applications.scala:528)
[error] dotty.tools.dotc.typer.Applications$Application.matchArgs(Applications.scala:593)
[error] dotty.tools.dotc.typer.Applications$Application.init(Applications.scala:431)
[error] dotty.tools.dotc.typer.Applications$TypedApply.<init>(Applications.scala:697)
[error] dotty.tools.dotc.typer.Applications$ApplyToUntyped.<init>(Applications.scala:801)
[error] dotty.tools.dotc.typer.Applications.ApplyTo(Applications.scala:1012)
[error] dotty.tools.dotc.typer.Applications.ApplyTo$(Applications.scala:301)
[error] dotty.tools.dotc.typer.Typer.ApplyTo(Typer.scala:102)
[error] dotty.tools.dotc.typer.Applications.simpleApply$1(Applications.scala:859)
[error] dotty.tools.dotc.typer.Applications.realApply$5$$anonfun$4(Applications.scala:938)
[error] dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:2852)
[error] dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:949)
[error] dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:987)
[error] dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:301)
[error] dotty.tools.dotc.typer.Inliner$InlineTyper.typedApply(Inliner.scala:1301)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2597)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2659)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2724)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2728)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2844)
[error] dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:834)
[error] dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:987)
[error] dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:301)
[error] dotty.tools.dotc.typer.Inliner$InlineTyper.typedApply(Inliner.scala:1301)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2597)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2659)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2724)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2728)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2844)
[error] dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:979)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2605)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2659)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2724)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2721)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2728)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2844)
[error] dotty.tools.dotc.typer.Inliner$InlineTyper.typedApply(Inliner.scala:1306)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2597)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2659)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2724)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2728)
[error] dotty.tools.dotc.typer.ReTyper.typedTyped(ReTyper.scala:62)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2602)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2659)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2724)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2721)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2728)
[error] dotty.tools.dotc.typer.Inliner.inlined(Inliner.scala:803)
[error] dotty.tools.dotc.typer.Inliner$.inlineCall(Inliner.scala:139)
[error] dotty.tools.dotc.transform.Inlining$InliningTreeMap.transform(Inlining.scala:86)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$1(Trees.scala:1422)
[error] scala.collection.immutable.List.mapConserve(List.scala:472)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1422)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformStats(Trees.scala:1420)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1335)
[error] dotty.tools.dotc.ast.TreeMapWithImplicits.transform(TreeMapWithImplicits.scala:92)
[error] dotty.tools.dotc.transform.Inlining$InliningTreeMap.transform(Inlining.scala:82)
[error] dotty.tools.dotc.ast.TreeMapWithImplicits.transform(TreeMapWithImplicits.scala:99)
[error] dotty.tools.dotc.transform.Inlining$InliningTreeMap.transform(Inlining.scala:80)
[error] dotty.tools.dotc.ast.TreeMapWithImplicits.traverse$1(TreeMapWithImplicits.scala:53)
[error] dotty.tools.dotc.ast.TreeMapWithImplicits.transformStats(TreeMapWithImplicits.scala:60)
[error] dotty.tools.dotc.ast.TreeMapWithImplicits.transform(TreeMapWithImplicits.scala:111)
[error] dotty.tools.dotc.transform.Inlining$InliningTreeMap.transform(Inlining.scala:80)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1398)
[error] dotty.tools.dotc.ast.TreeMapWithImplicits.transform(TreeMapWithImplicits.scala:104)
[error] dotty.tools.dotc.transform.Inlining$InliningTreeMap.transform(Inlining.scala:80)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$1(Trees.scala:1422)
[error] scala.collection.immutable.List.mapConserve(List.scala:472)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1422)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformStats(Trees.scala:1420)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1407)
[error] dotty.tools.dotc.ast.TreeMapWithImplicits.transform(TreeMapWithImplicits.scala:104)
[error] dotty.tools.dotc.transform.Inlining$InliningTreeMap.transform(Inlining.scala:94)
[error] dotty.tools.dotc.transform.Inlining$$anon$2.transform(Inlining.scala:72)
[error] dotty.tools.dotc.transform.MacroTransform.run(MacroTransform.scala:21)
[error] dotty.tools.dotc.transform.Inlining.run(Inlining.scala:44)
[error] dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:303)
[error] scala.collection.immutable.List.map(List.scala:246)
[error] dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:304)
[error] dotty.tools.dotc.transform.Inlining.runOn(Inlining.scala:48)
[error] dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:205)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error] dotty.tools.dotc.Run.runPhases$5(Run.scala:215)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:223)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:230)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:172)
[error] dotty.tools.dotc.Driver.finish(Driver.scala:60)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:40)
[error] dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
[error] dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:92)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:186)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:241)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:176)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:157)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:157)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:204)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:573)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:573)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:173)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:171)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:458)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:261)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:413)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:498)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:400)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:165)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:573)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:491)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2177)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2134)
[error] sbt.internal.io.Retry$.apply(Retry.scala:40)
[error] sbt.internal.io.Retry$.apply(Retry.scala:23)
[error] sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2130)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] sbt.Execute.work(Execute.scala:291)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
[error] java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
[error] java.base/java.lang.Thread.run(Thread.java:832)
[error]            
[error] java.lang.AssertionError: assertion failed
[error] 	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:11)
[error] 	at dotty.tools.dotc.typer.EtaExpansion$.etaExpand(EtaExpansion.scala:226)
[error] 	at dotty.tools.dotc.typer.Typer.adaptNoArgsUnappliedMethod$2(Typer.scala:3383)
[error] 	at dotty.tools.dotc.typer.Typer.adaptNoArgs$1(Typer.scala:3528)
[error] 	at dotty.tools.dotc.typer.Typer.adapt1(Typer.scala:3751)
[error] 	at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:3058)
[error] 	at dotty.tools.dotc.typer.ProtoTypes$FunProto.typedArg(ProtoTypes.scala:416)
[error] 	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:802)
[error] 	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:802)
[error] 	at dotty.tools.dotc.typer.Applications$Application.addTyped$1(Applications.scala:528)
[error] 	at dotty.tools.dotc.typer.Applications$Application.matchArgs(Applications.scala:593)
[error] 	at dotty.tools.dotc.typer.Applications$Application.init(Applications.scala:431)
[error] 	at dotty.tools.dotc.typer.Applications$TypedApply.<init>(Applications.scala:697)
[error] 	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.<init>(Applications.scala:801)
[error] 	at dotty.tools.dotc.typer.Applications.ApplyTo(Applications.scala:1012)
[error] 	at dotty.tools.dotc.typer.Applications.ApplyTo$(Applications.scala:301)
[error] 	at dotty.tools.dotc.typer.Typer.ApplyTo(Typer.scala:102)
[error] 	at dotty.tools.dotc.typer.Applications.simpleApply$1(Applications.scala:859)
[error] 	at dotty.tools.dotc.typer.Applications.realApply$5$$anonfun$4(Applications.scala:938)
[error] 	at dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:2852)
[error] 	at dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:949)
[error] 	at dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:987)
[error] 	at dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:301)
[error] 	at dotty.tools.dotc.typer.Inliner$InlineTyper.typedApply(Inliner.scala:1301)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2597)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2659)
[error] 	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
[error] 	at dotty.tools.dotc.typer.ProtoTypes$FunProto.$anonfun$2(ProtoTypes.scala:414)
[error] 	at dotty.tools.dotc.typer.ProtoTypes$FunProto.cacheTypedArg(ProtoTypes.scala:370)
[error] 	at dotty.tools.dotc.typer.ProtoTypes$FunProto.typedArg(ProtoTypes.scala:415)
[error] 	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:802)
[error] 	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:802)
[error] 	at dotty.tools.dotc.typer.Applications$Application.addTyped$1(Applications.scala:528)
[error] 	at dotty.tools.dotc.typer.Applications$Application.matchArgs(Applications.scala:593)
[error] 	at dotty.tools.dotc.typer.Applications$Application.init(Applications.scala:431)
[error] 	at dotty.tools.dotc.typer.Applications$TypedApply.<init>(Applications.scala:697)
[error] 	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.<init>(Applications.scala:801)
[error] 	at dotty.tools.dotc.typer.Applications.ApplyTo(Applications.scala:1012)
[error] 	at dotty.tools.dotc.typer.Applications.ApplyTo$(Applications.scala:301)
[error] 	at dotty.tools.dotc.typer.Typer.ApplyTo(Typer.scala:102)
[error] 	at dotty.tools.dotc.typer.Applications.simpleApply$1(Applications.scala:859)
[error] 	at dotty.tools.dotc.typer.Applications.realApply$5$$anonfun$4(Applications.scala:938)
[error] 	at dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:2852)
[error] 	at dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:949)
[error] 	at dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:987)
[error] 	at dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:301)
[error] 	at dotty.tools.dotc.typer.Inliner$InlineTyper.typedApply(Inliner.scala:1301)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2597)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2659)
[error] 	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2724)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2728)
[error] 	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2844)
[error] 	at dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:834)
[error] 	at dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:987)
[error] 	at dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:301)
[error] 	at dotty.tools.dotc.typer.Inliner$InlineTyper.typedApply(Inliner.scala:1301)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2597)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2659)
[error] 	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2724)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2728)
[error] 	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2844)
[error] 	at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:979)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2605)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2659)
[error] 	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2724)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2721)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2728)
[error] 	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2844)
[error] 	at dotty.tools.dotc.typer.Inliner$InlineTyper.typedApply(Inliner.scala:1306)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2597)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2659)
[error] 	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2724)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2728)
[error] 	at dotty.tools.dotc.typer.ReTyper.typedTyped(ReTyper.scala:62)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2602)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2659)
[error] 	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2724)
```
</details>

--- a/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala
+++ b/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala
@@ -592,11 +592,15 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
 
     object Apply extends ApplyModule:
       def apply(fun: Term, args: List[Term]): Apply =
+        yCheckArgs(args)
         withDefaultPos(tpd.Apply(fun, args))
       def copy(original: Tree)(fun: Term, args: List[Term]): Apply =
+        yCheckArgs(args)
         tpd.cpy.Apply(original)(fun, args)
       def unapply(x: Apply): (Term, List[Term]) =
         (x.fun, x.args)
+      private def yCheckArgs(args: List[Term]): Unit =
+        if yCheck then args.foreach(yCheckValidExpr)
     end Apply
 
     given ApplyMethods: ApplyMethods with
@@ -2873,6 +2877,12 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
           Extractors.showConstant(using QuotesImpl.this)(const)
 
     end Printer
+
+    private def yCheckValidExpr(term: Term): Unit =
+      if yCheck then
+        assert(!term.tpe.widenDealias.isInstanceOf[dotc.core.Types.MethodicType],
+          "Reference to a method must be eta-expanded before it is used as an expression: " + term.show)
+
   end reflect
 
   def unpickleExpr[T](pickled: String | List[String], typeHole: (Int, Seq[Any]) => scala.quoted.Type[?], termHole: (Int, Seq[Any], scala.quoted.Quotes) => scala.quoted.Expr[?]): scala.quoted.Expr[T] =

--- a/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala
+++ b/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala
@@ -264,7 +264,7 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
     object DefDef extends DefDefModule:
       def apply(symbol: Symbol, rhsFn: List[List[Tree]] => Option[Term]): DefDef =
         withDefaultPos(tpd.DefDef(symbol.asTerm, prefss =>
-          yCheckedOwners(rhsFn(prefss), symbol).getOrElse(tpd.EmptyTree)
+          yCheckedOwners(yCheckValidExpr(rhsFn(prefss)), symbol).getOrElse(tpd.EmptyTree)
         ))
       def copy(original: Tree)(name: String, paramss: List[ParamClause], tpt: TypeTree, rhs: Option[Term]): DefDef =
         tpd.cpy.DefDef(original)(name.toTermName, paramss, tpt, yCheckedOwners(rhs, original.symbol).getOrElse(tpd.EmptyTree))
@@ -293,9 +293,9 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
 
     object ValDef extends ValDefModule:
       def apply(symbol: Symbol, rhs: Option[Term]): ValDef =
-        tpd.ValDef(symbol.asTerm, yCheckedOwners(rhs, symbol).getOrElse(tpd.EmptyTree))
+        tpd.ValDef(symbol.asTerm, yCheckedOwners(yCheckValidExpr(rhs), symbol).getOrElse(tpd.EmptyTree))
       def copy(original: Tree)(name: String, tpt: TypeTree, rhs: Option[Term]): ValDef =
-        tpd.cpy.ValDef(original)(name.toTermName, tpt, yCheckedOwners(rhs, original.symbol).getOrElse(tpd.EmptyTree))
+        tpd.cpy.ValDef(original)(name.toTermName, tpt, yCheckedOwners(yCheckValidExpr(rhs), original.symbol).getOrElse(tpd.EmptyTree))
       def unapply(vdef: ValDef): (String, TypeTree, Option[Term]) =
         (vdef.name.toString, vdef.tpt, optional(vdef.rhs))
 
@@ -568,9 +568,9 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
 
     object NamedArg extends NamedArgModule:
       def apply(name: String, arg: Term): NamedArg =
-        withDefaultPos(tpd.NamedArg(name.toTermName, arg))
+        withDefaultPos(tpd.NamedArg(name.toTermName, yCheckValidExpr(arg)))
       def copy(original: Tree)(name: String, arg: Term): NamedArg =
-        tpd.cpy.NamedArg(original)(name.toTermName, arg)
+        tpd.cpy.NamedArg(original)(name.toTermName, yCheckValidExpr(arg))
       def unapply(x: NamedArg): (String, Term) =
         (x.name.toString, x.value)
     end NamedArg
@@ -592,15 +592,13 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
 
     object Apply extends ApplyModule:
       def apply(fun: Term, args: List[Term]): Apply =
-        yCheckArgs(args)
+        yCheckValidExprs(args)
         withDefaultPos(tpd.Apply(fun, args))
       def copy(original: Tree)(fun: Term, args: List[Term]): Apply =
-        yCheckArgs(args)
+        yCheckValidExprs(args)
         tpd.cpy.Apply(original)(fun, args)
       def unapply(x: Apply): (Term, List[Term]) =
         (x.fun, x.args)
-      private def yCheckArgs(args: List[Term]): Unit =
-        if yCheck then args.foreach(yCheckValidExpr)
     end Apply
 
     given ApplyMethods: ApplyMethods with
@@ -669,9 +667,9 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
 
     object Typed extends TypedModule:
       def apply(expr: Term, tpt: TypeTree): Typed =
-        withDefaultPos(tpd.Typed(expr, tpt))
+        withDefaultPos(tpd.Typed(yCheckValidExpr(expr), tpt))
       def copy(original: Tree)(expr: Term, tpt: TypeTree): Typed =
-        tpd.cpy.Typed(original)(expr, tpt)
+        tpd.cpy.Typed(original)(yCheckValidExpr(expr), tpt)
       def unapply(x: Typed): (Term, TypeTree) =
         (x.expr, x.tpt)
     end Typed
@@ -693,9 +691,9 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
 
     object Assign extends AssignModule:
       def apply(lhs: Term, rhs: Term): Assign =
-        withDefaultPos(tpd.Assign(lhs, rhs))
+        withDefaultPos(tpd.Assign(lhs, yCheckValidExpr(rhs)))
       def copy(original: Tree)(lhs: Term, rhs: Term): Assign =
-        tpd.cpy.Assign(original)(lhs, rhs)
+        tpd.cpy.Assign(original)(lhs, yCheckValidExpr(rhs))
       def unapply(x: Assign): (Term, Term) =
         (x.lhs, x.rhs)
     end Assign
@@ -758,7 +756,7 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
     object Lambda extends LambdaModule:
       def apply(owner: Symbol, tpe: MethodType, rhsFn: (Symbol, List[Tree]) => Tree): Block =
         val meth = dotc.core.Symbols.newSymbol(owner, nme.ANON_FUN, Synthetic | Method, tpe)
-        tpd.Closure(meth, tss => yCheckedOwners(rhsFn(meth, tss.head.map(withDefaultPos)), meth))
+        tpd.Closure(meth, tss => yCheckedOwners(yCheckValidExpr(rhsFn(meth, tss.head.map(withDefaultPos))), meth))
 
       def unapply(tree: Block): Option[(List[ValDef], Term)] = tree match {
         case Block((ddef @ DefDef(_, TermParamClause(params) :: Nil, _, Some(body))) :: Nil, Closure(meth, _))
@@ -778,9 +776,9 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
 
     object If extends IfModule:
       def apply(cond: Term, thenp: Term, elsep: Term): If =
-        withDefaultPos(tpd.If(cond, thenp, elsep))
+        withDefaultPos(tpd.If(yCheckValidExpr(cond), yCheckValidExpr(thenp), yCheckValidExpr(elsep)))
       def copy(original: Tree)(cond: Term, thenp: Term, elsep: Term): If =
-        tpd.cpy.If(original)(cond, thenp, elsep)
+        tpd.cpy.If(original)(yCheckValidExpr(cond), yCheckValidExpr(thenp), yCheckValidExpr(elsep))
       def unapply(tree: If): (Term, Term, Term) =
         (tree.cond, tree.thenp, tree.elsep)
     end If
@@ -804,10 +802,10 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
 
     object Match extends MatchModule:
       def apply(selector: Term, cases: List[CaseDef]): Match =
-        withDefaultPos(tpd.Match(selector, cases))
+        withDefaultPos(tpd.Match(yCheckValidExpr(selector), cases))
 
       def copy(original: Tree)(selector: Term, cases: List[CaseDef]): Match =
-        tpd.cpy.Match(original)(selector, cases)
+        tpd.cpy.Match(original)(yCheckValidExpr(selector), cases)
 
       def unapply(x: Match): (Term, List[CaseDef]) =
         (x.scrutinee, x.cases)
@@ -854,9 +852,9 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
 
     object Try extends TryModule:
       def apply(expr: Term, cases: List[CaseDef], finalizer: Option[Term]): Try =
-        withDefaultPos(tpd.Try(expr, cases, finalizer.getOrElse(tpd.EmptyTree)))
+        withDefaultPos(tpd.Try(yCheckValidExpr(expr), cases, finalizer.getOrElse(tpd.EmptyTree)))
       def copy(original: Tree)(expr: Term, cases: List[CaseDef], finalizer: Option[Term]): Try =
-        tpd.cpy.Try(original)(expr, cases, finalizer.getOrElse(tpd.EmptyTree))
+        tpd.cpy.Try(original)(yCheckValidExpr(expr), cases, finalizer.getOrElse(tpd.EmptyTree))
       def unapply(x: Try): (Term, List[CaseDef], Option[Term]) =
         (x.body, x.cases, optional(x.finalizer))
     end Try
@@ -879,9 +877,9 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
 
     object Return extends ReturnModule:
       def apply(expr: Term, from: Symbol): Return =
-        withDefaultPos(tpd.Return(expr, from))
+        withDefaultPos(tpd.Return(yCheckValidExpr(expr), from))
       def copy(original: Tree)(expr: Term, from: Symbol): Return =
-        tpd.cpy.Return(original)(expr, tpd.ref(from))
+        tpd.cpy.Return(original)(yCheckValidExpr(expr), tpd.ref(from))
       def unapply(x: Return): (Term, Symbol) =
         (x.expr, x.from.symbol)
     end Return
@@ -903,8 +901,10 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
 
     object Repeated extends RepeatedModule:
       def apply(elems: List[Term], elemtpt: TypeTree): Repeated =
+        yCheckValidExprs(elems)
         withDefaultPos(tpd.SeqLiteral(elems, elemtpt))
       def copy(original: Tree)(elems: List[Term], elemtpt: TypeTree): Repeated =
+        yCheckValidExprs(elems)
         tpd.cpy.SeqLiteral(original)(elems, elemtpt)
       def unapply(x: Repeated): (List[Term], TypeTree) =
         (x.elems, x.elemtpt)
@@ -927,9 +927,9 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
 
     object Inlined extends InlinedModule:
       def apply(call: Option[Tree], bindings: List[Definition], expansion: Term): Inlined =
-        withDefaultPos(tpd.Inlined(call.getOrElse(tpd.EmptyTree), bindings.map { case b: tpd.MemberDef => b }, expansion))
+        withDefaultPos(tpd.Inlined(call.getOrElse(tpd.EmptyTree), bindings.map { case b: tpd.MemberDef => b }, yCheckValidExpr(expansion)))
       def copy(original: Tree)(call: Option[Tree], bindings: List[Definition], expansion: Term): Inlined =
-        tpd.cpy.Inlined(original)(call.getOrElse(tpd.EmptyTree), bindings.asInstanceOf[List[tpd.MemberDef]], expansion)
+        tpd.cpy.Inlined(original)(call.getOrElse(tpd.EmptyTree), bindings.asInstanceOf[List[tpd.MemberDef]], yCheckValidExpr(expansion))
       def unapply(x: Inlined): (Option[Tree /* Term | TypeTree */], List[Definition], Term) =
         (optional(x.call), x.bindings, x.body)
     end Inlined
@@ -982,9 +982,9 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
 
     object While extends WhileModule:
       def apply(cond: Term, body: Term): While =
-        withDefaultPos(tpd.WhileDo(cond, body))
+        withDefaultPos(tpd.WhileDo(yCheckValidExpr(cond), yCheckValidExpr(body)))
       def copy(original: Tree)(cond: Term, body: Term): While =
-        tpd.cpy.WhileDo(original)(cond, body)
+        tpd.cpy.WhileDo(original)(yCheckValidExpr(cond), yCheckValidExpr(body))
       def unapply(x: While): (Term, Term) =
         (x.cond, x.body)
     end While
@@ -2834,6 +2834,18 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
             case _ => traverseChildren(t)
       }.traverse(tree)
 
+    private def yCheckValidExprs(terms: List[Term]): terms.type =
+      if yCheck then terms.foreach(yCheckValidExpr)
+      terms
+    private def yCheckValidExpr(termOpt: Option[Term]): termOpt.type =
+      if yCheck then termOpt.foreach(yCheckValidExpr)
+      termOpt
+    private def yCheckValidExpr(term: Term): term.type =
+      if yCheck then
+        assert(!term.tpe.widenDealias.isInstanceOf[dotc.core.Types.MethodicType],
+          "Reference to a method must be eta-expanded before it is used as an expression: " + term.show)
+      term
+
     object Printer extends PrinterModule:
 
       lazy val TreeCode: Printer[Tree] = new Printer[Tree]:
@@ -2877,11 +2889,6 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
           Extractors.showConstant(using QuotesImpl.this)(const)
 
     end Printer
-
-    private def yCheckValidExpr(term: Term): Unit =
-      if yCheck then
-        assert(!term.tpe.widenDealias.isInstanceOf[dotc.core.Types.MethodicType],
-          "Reference to a method must be eta-expanded before it is used as an expression: " + term.show)
 
   end reflect
 

--- a/tests/neg-macros/i11483/Macro_1.scala
+++ b/tests/neg-macros/i11483/Macro_1.scala
@@ -1,0 +1,59 @@
+package x
+
+import scala.annotation._
+import scala.quoted._
+import scala.compiletime._
+
+
+trait CpsMonad[F[_]]:
+
+  def pure[T](x:T):F[T]
+
+  def impure[T](x:F[T]):T
+
+  def map[A,B](x:F[A])(f: A=>B):F[B]
+
+
+@compileTimeOnly("await should be inside async block")
+def await[F[_],T](f:F[T])(using am:CpsMonad[F]):T = ???
+
+inline given conversion[F[_],T](using CpsMonad[F]): Conversion[F[T],T] =
+           x => await(x)
+
+
+object X {
+
+ inline def process[F[_], T](inline t:T)(using m: CpsMonad[F]):F[T] =
+    ${ processImpl[F,T]('t, 'm) }
+
+
+ def processImpl[F[_]:Type, T:Type](t:Expr[T], m:Expr[CpsMonad[F]])(using Quotes):Expr[F[T]] =
+   import quotes.reflect._
+   val r = processTree[F,T](t.asTerm, m.asTerm)
+   r.asExprOf[F[T]]
+
+
+ def processTree[F[_]:Type, T:Type](using Quotes)(t: quotes.reflect.Term, m: quotes.reflect.Term):quotes.reflect.Term =
+   import quotes.reflect._
+   val r: Term = t match
+     case Inlined(_,List(),body) => processTree[F,T](body, m)
+     case Inlined(d,bindings,body) =>
+       Inlined(d,bindings,processTree[F,T](body, m))
+     case Block(stats,expr) => Block(stats,processTree(expr, m))
+     case Apply(Apply(TypeApply(Ident("await"),targs),List(body)),List(m)) => body
+     case Apply(f,List(arg)) =>
+          val nArg = processTree[F,String](arg, m)
+          Apply(Apply(TypeApply(Select.unique(m,"map"),
+                               List(Inferred(arg.tpe.widen),Inferred(t.tpe.widen))
+                      ),
+                      List(nArg)),
+                 List(f)
+          )
+     case Apply(f,List()) =>
+          Apply(TypeApply(Select.unique(m,"pure"),List(Inferred(t.tpe.widen))),List(t))
+     case Typed(x,tp) => Typed(processTree(x,m), Inferred(TypeRepr.of[F].appliedTo(tp.tpe)) )
+     case _ => throw new RuntimeException(s"tree not recoginized: $t")
+   r
+
+
+}

--- a/tests/neg-macros/i11483/Test_2.scala
+++ b/tests/neg-macros/i11483/Test_2.scala
@@ -1,0 +1,26 @@
+package x
+
+import scala.language.implicitConversions
+import scala.concurrent.Future
+
+given FutureAsyncMonad: CpsMonad[Future] with
+  def pure[T](t:T): Future[T] = ???
+  def impure[T](t:Future[T]): T = ???
+  def map[A,B](x:Future[A])(f: A=>B): Future[B] = ???
+
+
+object Api:
+
+  def doSomething():Future[String] =
+    Future.successful("doSomething")
+
+  def println(x:String):Unit =
+    Console.println(x)
+
+
+object Main:
+
+  def main(args: Array[String]): Unit =
+    X.process[Future,Unit]{ // error
+      Api.println(Api.doSomething())
+    }

